scriptmodules/ports/jfsw.sh

```diff
@@ -0,0 +1,64 @@
+#!/usr/bin/env bash
+
+# This file is part of The RetroPie Project
+#
+# The RetroPie Project is the legal property of its developers, whose names are
+# too numerous to list here. Please refer to the COPYRIGHT.md file distributed with this source.
+#
+# See the LICENSE.md file at the top-level directory of this distribution and
+# at https://raw.githubusercontent.com/RetroPie/RetroPie-Setup/master/LICENSE.md
+#
+
+rp_module_id="jfsw"
+rp_module_desc="Shadow Warrior source port by JonoF"
+rp_module_help="Place your registered version game files in $romdir/ports/shadowwarrior"
+rp_module_licence="GPL2 https://raw.githubusercontent.com/jonof/jfsw/master/GPL.TXT"
+rp_module_repo="git https://github.com/jonof/jfsw.git master"
+rp_module_section="exp"
+rp_module_flags=""
+
+function depends_jfsw() {
+    depends_jfduke3d
+}
+
+function sources_jfsw() {
+    gitPullOrClone
+}
+
+function build_jfsw() {
+    build_jfduke3d "shadowwarrior" "sw"
+}
+
+function install_jfsw() {
+    install_jfduke3d 'sw'
+}
+
+function gamedata_jfsw() {
+    local dest="$romdir/ports/shadowwarrior"
+    if [[ ! -n $(find $dest -maxdepth 1 -iname sw.grp) ]]; then
+        mkUserDir "$dest"
+        local temp="$(mktemp -d)"
+        download "ftp://ftp.3drealms.com/share/3dsw12.zip" "$temp"
+        unzip -L -o "$temp/3dsw12.zip" -d "$temp" swsw12.shr
+        unzip -L -o "$temp/swsw12.shr" -d "$dest" sw.grp sw.rts
+        rm -rf "$temp"
+    fi
+    chown -R $user:$user "$dest"
+}
+
+function configure_jfsw() {
+    local gamedir="$romdir/ports/shadowwarrior"
+
+    mkRomDir "ports/shadowwarrior"
+    moveConfigDir "$home/.jfsw" "$md_conf_root/sw/jfsw"
```

Comment:
Here, the opposite. Should this just use `configs/ports/sw` like the RP-Extra module, or why the extra layer?

Reply:
I though the same, but `configs/ports/duke3d` points ***directly*** to the config to EDuke32, and your idea would force the config of JFDuke3D to go inside the config of EDuke32 as a subdirectory. Imagine how that would appear inside the config directories of RetroPie (`/opt/retropie/configs/ports`), and most importantly, how the "home" directory would appear:
```
ports
  |
  |- duke3d (1)
        |
        |- jfduke3d (2)

~
|
|- .config
|     |
|     |- eduke32 (1)
|          |
|          |- jfduke3d (2)
|
|- .jfduke3d (2)
```
`(1)` and `(2)` are to see where the links point.

The "_extra layer_" you mention was inspired on how is currently treated for _Quake 1_ and _2_, where the first layer is the name of the game, and inside them is every individual port for that game (compare `tyrquake.sh` and `darkplaces-quake.sh` to understand). It's just better organized and future-proofed, compared with how _Duke 3D_ does it, which just has "the name of the game" as the destination for the config of just one of the ports. It's just something that didn't age well, because of the need for organization when having more than one port for a game.

I don't find that joining all the configs on the different ports under the "first port made" is a good idea, especially if the user wants to back up a config for a certain port; this user might find every single config for every port of that game included. Or if the players want to "reset the config" by cleaning the directory, they will find that they deleted the configs for all of the ports of that game.

Of course, there may be a better solution, but I dealed with a legacy problem the best I could. I just don't find mixing directories is a good idea, but if the owners of the project want that, it'll be done.

Reply:
So what you're saying is that it *should* actually be `configs/ports/duke3d/eduke32`, but that that ship has sailed...

Reply:
Yeah, exactly, that's what EDuke32 should have used since the beginning. Although I understand no one would have predicted the use of different ports for the same game, it would have been nice if it was considered in some form. In the ideal world, it would have been great if this change could be done without the user having to do anything on their end.

Still, this had me thinking again. What if JFDuke3D, instead of using `configs/ports/jfduke3d`, used `configs/ports/dukenukem3d/jfduke3d`? Even if we have the same game twice, at least there would be a place where the configs of a potential 3rd port of _Duke Nukem 3D_ could be located, without filling more places on the "root" of `ports`. If you have a better idea, I'm all ears.

Reply:
Thanks. The config locations for some things are not ideal, due to later additions. It might be worth re-organising the configs for eduke32. I will come back to you. Sorry for the delay.